[MIG][13.0] stock_orderpoint_route: Migration to 13.0

guewen opened the pull request:
Port #785, #537

JordiBForgeFlow committed:
[11.][ADD] stock_orderpoint_route

Commit:
Migrate stock_orderpoint_route to 12.0

Commit:
Use SavepointCase in tests

To avoid executing the setup for each test

Commit:
[IMP] stock_orderpoint_route: black, isort

Commit:
[MIG] stock_orderpoint_route: Migration to 13.0

simahawk commented on stock_orderpoint_route/models/stock_warehouse_orderpoint.py:
```diff
@@ -0,0 +1,54 @@
+        route_obj = self.env["stock.location.route"]
+        for wh in self.mapped("warehouse_id"):
+            wh_routes = wh.route_ids
+            for record in self.filtered(lambda r: r.warehouse_id == wh):
```
can we avoid double loop?
can we avoid browsing?

guewen replied:
I ported the module as is, note: the number of warehouses is generally very low, shouldn't have too much an impact

guewen replied:
Fixed, we have less queries with a single, simple loop.

simahawk commented on stock_orderpoint_route/models/stock_warehouse_orderpoint.py:
```diff
@@ -0,0 +1,54 @@
+                if record.warehouse_id:
+                    routes |= wh_routes
+                parents = record.get_parents()
+                record.route_ids = routes.filtered(
```
can we split this to `_get_routes_whatever`?

simahawk commented on stock_orderpoint_route/models/stock_warehouse_orderpoint.py:
```diff
@@ -0,0 +1,54 @@
+                    )
+                )
+
+    def get_parents(self):
```
can't this be solved using parent_path?

simahawk commented on stock_orderpoint_route/tests/test_stock_orderpoint_route.py:
```diff
@@ -0,0 +1,176 @@
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+
```
set tracking disable pls

simahawk commented on stock_orderpoint_route/tests/test_stock_orderpoint_route.py:
```diff
@@ -0,0 +1,176 @@
+    def _create_user(cls, name, group_ids, company_ids):
+        return (
+            cls.env["res.users"]
+            .with_context({"no_reset_password": True})
```
use kw arg to not override ctx completely

guewen replied:
```suggestion
            .with_context(no_reset_password=True)
```

Commit:
run pre-commit with new prettiers

Commit:
Improve code after reviews

* performance (less queries executed)
* disable tracking in tests (faster)
* use with_user instead of sudo
* the double for loop in _compute_route_ids actually generates one more
  query as the simple one in this commit (thanks to the cache)
* extract a method
* the parent locations can be found using 'parent_of' which will use the
  parent_path under the hood (example: 1/7/8 will return locations 1, 7,
  8)

guewen commented:
@simahawk rebased, fixed your comments in the last commit

simahawk commented:
@JordiBForgeFlow could you pass by for a review? :)

jgrandguillaume reviewed (approved):
Functional review is ok on my side

simahawk commented:
/ocabot merge
